### PR TITLE
[Orders with Coupons M6] Collapsible product row component

### LINKED ISSUE #10833: [Orders with Coupons M6] Update product rows with `CollapsibleProductRow` component

### PR DESCRIPTION
**Pull request by iamgabrielma**
Closes: #10833 

## Description
With this PR we add different SwiftUI components to make the product rows for Order Creation collapsable: `CollapsibleProductRowCard`, `CollapsibleProductCardPriceSummary`, and `SimplifiedProductRow`

The design is not yet as the [proposed designs](https://www.figma.com/file/FTOWxedKUyDbBdOrxRfMyI/Coupons-in-Order-Creation?type=design&node-id=14-9617&mode=design&t=gCbDARMVXEBqwarn-0), since we have to add further components and polish the details, but gets as close as possible. Some items that still needs addressing:
* The chevron for collapsing/expanding is centered due to reusing existing components, we'll duplicate this in a future PR and adapt it properly.
* There's no discounts-related functionality or design, this will be tackled on a different PR [here](https://github.com/woocommerce/woocommerce-ios/issues/10837).

CollapsibleProductRow

## Testing instructions
* Switch `ordersWithCouponsM6` feature flag to `true`
* Go to `Orders` > `+` > `+ Add Products` > Add different products to your order: 
  * Digital products in stock.
  * Physical products in stock where "manage stock" option is enabled, and we've given a stock quantity.

Observe that:
* Each product added to the order is collapsable, and its state is independent from other products.
* Placeholder image, product title, stock, price/quantity breakdown, and total appears and is updated when stock changes.
* Upon collapsing the view price and quantity appears, and upon expanding the view this changes to the product SKU. The border color changes as well when a product card is collapsed/expanded.
* The quantity stepper behaviour hasn't changed (this is finicky if we increase/decrease fast, but it's an underlying issue): 
  * Tapping on `+` in the quantity stepper increases product count.
  * Tapping on `-` decreases it.
  * Tapping on `-` and reaching zero removes the product from the order.
* Tapping on `Remove Product from Order` currently does nothing at the moment. We'll implement it on a [separate PR](https://github.com/woocommerce/woocommerce-ios/issues/10834).
* Taping `Create` creates the Order normally, with the expected products and quantities.

### CODE REVIEW

**Comment by iamgabrielma** [comment]
The failing UI test is unrelated to the PR and seems to be currently in trunk. Reported here: p1696411510426629-slack-C6H8C3G23. Update: Resolved.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+struct CollapsibleProductRowCard: View {
+    @ObservedObject var viewModel: ProductRowViewModel
+    @State private var isCollapsed: Bool = true
+
+    private var isExpanded: Binding<Bool> {
+        Binding<Bool>(
+            get: { !self.isCollapsed },
+            set: { self.isCollapsed = !$0 }
+        )
+    }
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: Happy to hear if you have a better suggestion for this bit of logic, ideally I wanted to use `!isCollapsed` when passing the property down to the `CollapsibleView.shouldShowDividers`, however since uses a Binding<Bool> type for 2-way communication I haven't found a better way to do it.

Since I plan to adapt the `CollapsibleView` to our use case specifically in a future PR to get closer to the proposed design is not a big deal either, as most likely we'll get rid of this shortly.

Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: So... to solve this immediate issue, you could make `shouldShowDividers` a private computed var:
```
    private var shouldShowDividers: Bool {
        return !isCollapsed
    }
```

Which seems to match the current behaviour.

I'd suggest though... perhap `CollapsibleView` is doing too much? It's generically named, and could be generically useful, but I don't think we'll want the divider behaviour everywhere, and it won't neccesarily be a label when collapsed.

Instead, you could try keeping two ViewBuilders, but making them `collapsedView` and `expandedView` – where collapsed is a subset of expanded, and the `CollapsibleView` shows one at a time, rather than one when collapsed and two when expanded.

That way the dividers can be defined in the expanded view only, and you don't need this logic.

Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: I wrote the above thinking that you'd added CollapsibleView in this PR; I see it's pre-existing. I still think all this applies, but it does depend where/how much we're already using CollapsibleView elsewhere.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    init(viewModel: ProductRowViewModel) {
+        self.viewModel = viewModel
+    }
+
+    var body: some View {
+        HStack {
+            HStack {
+                Text(viewModel.quantity.formatted())
+                    .foregroundColor(.gray)
+                Text("x")
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: This should probably be a multiplication symbol (×) rather than a letter x

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        static let frameCornerRadius: CGFloat = 4
+        static let borderLineWidth: CGFloat = 1
+    }
+
+    enum Localization {
+        static let priceLabel = NSLocalizedString(
+            "Price",
+            comment: "Text in the product row card that indicating the price of the product")
+        static let removeProductLabel = NSLocalizedString(
+            "Remove Product from order",
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]:
```suggestion
            "Remove product from order",
```

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift [hunk 4]
```diff
@@ -67,9 +67,34 @@ final class ProductRowViewModel: ObservableObject, Identifiable {
...
+        return currencyFormatter.formatAmount(price)
+    }
+
+    /// Formatted price label from multiplying product's price and quantity.
+    ///
+    var priceBeforeDiscountsLabel: String? {
+        guard let price = price else {
+            return nil
+        }
+        let productSubtotal = quantity * (currencyFormatter.convertToDecimal(price)?.decimalValue ?? Decimal.zero)
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift:91]: I don't think that defaulting to zero is appropriate here – it should probably be a `-` or similar non-number

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift [hunk 5]
```diff
@@ -324,6 +349,24 @@ final class ProductRowViewModel: ObservableObject, Identifiable {
...
+                let localizedStockQuantity = NumberFormatter.localizedString(from: stockQuantity as NSDecimalNumber, number: .decimal)
+                return String.localizedStringWithFormat(Localization.stockFormat, localizedStockQuantity)
+            } else {
+                return "In Stock"
+            }
+        case .outOfStock:
+            return "Out of Stock"
+        default:
+            return ""
+        }
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift]: Perhaps using the `description`, which is implemented on the model, would be better here?

**Review by joshheald — APPROVED** [review]
Nice one! A couple of suggestions but I didn't go too deep into the design etc since you're planning to do further work on it anyway.

Works as described

**Comment by joshheald** [comment]
> Works as described

One note post-review – it seems like if you edit an order and remove products (with the minus button) and tap done, further edits to that order always(?) fail:

```
⛔️ Error syncing order remotely: Dotcom Error: [woocommerce_rest_invalid_item_id] Order item ID provided is not associated with order.
```

**Comment by iamgabrielma** [comment]
> > Works as described
> 
> One note post-review – it seems like if you edit an order and remove products (with the minus button) and tap done, further edits to that order always(?) fail:
> 
> ```
> ⛔️ Error syncing order remotely: Dotcom Error: [woocommerce_rest_invalid_item_id] Order item ID provided is not associated with order.
> ```

Good catch, I was able to reproduce the same behaviour in trunk with the previous implementation, so I logged it here for the moment:

https://github.com/woocommerce/woocommerce-ios/issues/10874

~~I can't seem to see the sync error in the console~~ (Edit: I do see the error if I attempt to increase the quantity for a product that is supposed to be removed), but I see the same failing behaviour upon attempting to save the changes. I suspect could be related to using product row ID's in order to remove these from the order that is being edited, since these doesn't necessarily have to march the order item ID that we'll be sending to the remote, but I haven't dig deeper into it yet.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: The computed property makes it much cleaner, since it failed for me to pass the `!$isCollapsible` binding directly to the initializer, I didn't think about this solution. Thanks! Changed here: 793191caf8d4197bd4418e4ffa94241a9b841ca4

> I wrote the above thinking that you'd added CollapsibleView in this PR; I see it's pre-existing. I still think all this applies, but it does depend where/how much we're already using CollapsibleView elsewhere.

Agree, when revising the design for the collapsible component it might be worth to take this approach

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift (hunk 2 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: Changed here: f68395748ed52661fb08d4a24022eef10673f7ac

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift (hunk 3 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift]: ecd1c5245f8282f1039f383995cc08ca1bb62989

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift (hunk 4 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift:91]: This is something that felt odd as well: This bit it's copied from another method that performs the same logic, but then add discounts when needed, since I wanted to keep them separate. When testing what would happen if we pass a non-numeric-value, the currency formatter does return zero in any case, which I don't think is what we want in all cases, and feels a bit like counting on a hidden side effect in the currency formatter where we assume that a valid non-nil correct price will be passed into the method, and this will be correctly formatted, otherwise "it surely must be zero".

I opened a separate issue here: https://github.com/woocommerce/woocommerce-ios/issues/10875 so I can test further and improve it.

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift (hunk 5 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift]: Yes! 94c3357a538f77872af2ca7473d9d0885c24fa10

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift (hunk 4 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductRowViewModel.swift:91]: I forgot to add, I added additional Unit Tests to the currency formatter for this method to show that both zero and non-numeric-values are converted: 1a48e053754e8868b7f127c63823b2f419f4f644 so it can help to change the behaviour later.
